File: 2695-find-score-of-an-array-after-marking-all-elements/2695-find-score-of-an-array-after-marking-all-elements.cpp
class Solution {
public:
    long long findScore(vector<int>& nums) {
        int n = nums.size();
        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       greater<pair<int, int>>>
            pq;

        vector<int> visited(n, 0);

        for (int i = 0; i < n; i++)
            pq.push({nums[i], i});

        long long ans = 0;

        while (!pq.empty()) {
            auto it = pq.top();
            int element = it.first;
            int index = it.second;
            pq.pop();

            if (visited[index] == 1)
                continue;
            visited[index] = 1;

            ans += element;

            if (index - 1 >= 0)
                visited[index - 1] = 1;
            if (index + 1 < n)
                visited[index + 1] = 1;
        }

        return ans;
    }
};

File: 2695-find-score-of-an-array-after-marking-all-elements/2695-find-score-of-an-array-after-marking-all-elements.java
class Solution {
    public long findScore(int[] nums) {
        int n = nums.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) ->a[0]!=b[0]?a[0] - b[0]:a[1]-b[1]);
        for (int i = 0; i < n; i++) {
            pq.offer(new int[]{nums[i], i});
        }

        long sum = 0;
        boolean[] vis = new boolean[n];
        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            int val = top[0], ind = top[1];
            if (vis[ind]) continue;
            sum += val;
            vis[ind] = true;
            if (ind - 1 >= 0) vis[ind - 1] = true;
            if (ind + 1 < n) vis[ind + 1] = true;
        }
        return sum;
    }
}